Remove `couch_replicator_manager`

nickva opened the pull request:
Its logic is now spread amongst a few modules:
- `couch_multidb_changes`:
  Monitor `*/_replicator` dbs, run callbacks for db / doc changes.
- `couch_replicator_clustering`:
  Keep track of cluster membership. Trigger rescans, by restarting
  `couch_multidb_changes`. Resolve replication ownership.
- `couch_replicator`:
  Create local `_replicator` db on start-up. Add / remove replication jobs.
  Call `couch_replicator_docs` to write info back to replicator docs. Handle
  `_replicate` endpoint replications but use job scheduler instead.

Also:
- Dialyze more code, especially in couch_replicatorl and couch_replicator_docs.
- Keep `couch_replicator_manager` module around for a while in order to
  proxy calls made from other applications. Otherwise would have to clone other
  repos. Remove it and update code accordingly on final cleanup.
- Update `owner` function from couch_replicator_clustering to have result which
  makes more sense for replication jobs.
- Avoided touching scheduler as there are prs pending for it. Created a few
  job manipulation functions in couch_replicator.erl. Moving those to scheduler
  API eventually, as they access internal job scheduler data in a hacky way.
- A major change to behavior is errors are not written back to docs anymore
  (except errors in parsing the doc). Triggered updates are still written because
  they also tag the doc with the replication_id.

Updated eunit tests, all pass (still):

```
$ make eunit apps=couch_replicator
...
 [done in 51.177 s]
=======================================================
  All 171 tests passed.
```

Replications running from dev cluster with `rep` tool still trigger documents
and replicate data.

```
>>> rep.replicate_1_to_n_then_check_replication(2)
creating rdyno_src_0001
 > created  1 dbs with prefix rdyno_src_
creating rdyno_tgt_0001
creating rdyno_tgt_0002
 > created  2 dbs with prefix rdyno_tgt_
updating documents
 > _replicator rdyno_0001_0001 : 1-34f09a6a62181f4d4631bf8544b582fc
 > _replicator rdyno_0001_0002 : 1-0f52a360f959fea67e6cb20aa1bb0606
waiting for replication documents to trigger
 > function wait_to_trigger succeded after 0.011 +/- 10.0  sec.
all replication documents triggered
>>> update cycle 0  <<<
 > waiting for target  1
 > retrying function wait_till_dbs_equal
...
 > function wait_till_dbs_equal succeded after 49.574 +/- 1.0  sec.
 > waiting for target  2
 > function wait_till_dbs_equal succeded after 0.019 +/- 1.0  sec.
 > waiting to propagate changes from  1 to 2  : 49.623 sec.
```

nickva committed:
Remove `couch_replicator_manager`

Its logic is now spread amongst a few modules:

 - `couch_multidb_changes`:
    Monitor `*/_replicator` dbs, run callbacks for db / doc changes.

 - `couch_replicator_clustering`:
    Keep track of cluster membership. Trigger rescans, by restarting
    `couch_multidb_changes`. Resolve replication ownership.

 - `couch_replicator`:
    Create local `_replicator` db on start-up. Add / remove replication jobs.
    Call `couch_replicator_docs` to write info back to replicator docs. Handle
    `_replicate` endpoint replications but use job scheduler instead.

Also:

* Dialyze more code, especially in couch_replicatorl and couch_replicator_docs.

* Keep `couch_replicator_manager` module around for a while in order to
proxy calls made from other applications. Otherwise would have to clone other
repos. Remove it and update code accordingly on final cleanup.

* Update `owner` function from couch_replicator_clustering to have result which
makes more sense for replication jobs.

* Avoided touching scheduler as there are prs pending for it. Created a few
job manipulation functions in couch_replicator.erl. Moving those to scheduler
API eventually, as they access internal job scheduler data in a hacky way.

* A major change to behavior is errors are not written back to docs anymore
(except errors in parsing the doc). Triggered updates are still written because
they also tag the doc with the replication_id.

Updated eunit tests, all pass (still):

```
$ make eunit apps=couch_replicator
...
 [done in 51.177 s]
=======================================================
  All 171 tests passed.
```

Replications running from dev cluster with `rep` tool still trigger documents
and replicate data.

```
>>> rep.replicate_1_to_n_then_check_replication(2)
creating rdyno_src_0001
 > created  1 dbs with prefix rdyno_src_
creating rdyno_tgt_0001
creating rdyno_tgt_0002
 > created  2 dbs with prefix rdyno_tgt_
updating documents
 > _replicator rdyno_0001_0001 : 1-34f09a6a62181f4d4631bf8544b582fc
 > _replicator rdyno_0001_0002 : 1-0f52a360f959fea67e6cb20aa1bb0606
waiting for replication documents to trigger
 > function wait_to_trigger succeded after 0.011 +/- 10.0  sec.
all replication documents triggered
>>> update cycle 0  <<<
 > waiting for target  1
 > retrying function wait_till_dbs_equal
...
 > function wait_till_dbs_equal succeded after 49.574 +/- 1.0  sec.
 > waiting for target  2
 > function wait_till_dbs_equal succeded after 0.019 +/- 1.0  sec.
 > waiting to propagate changes from  1 to 2  : 49.623 sec.
```

sagelywizard commented on src/couch_replicator_scheduler_job.erl:
```diff
@@ -40,6 +40,15 @@
+
+%% behavior api
+-callback replication_started(Rep :: #rep{}) ->
+   ok.
```
Nit: Spacing is a little off

sagelywizard commented on src/couch_replicator_sup.erl:
```diff
@@ -13,43 +13,60 @@
             infinity,
             supervisor,
             [couch_replicator_scheduler_sup]},
+        {couch_replicator_scheduler,
```
Why'd you move this?

nickva replied:
Because it is a rest_for_one supervisor. The scheduler_sup should be up before scheduler itself so it can receive calls from replicator scheduler. For example it calls `couch_replicator_scheduler_sup:start_child()` so _sup should be up by then.

sagelywizard commented on src/couch_replicator_sup.erl:
```diff
@@ -13,43 +13,60 @@
+            [couch_multidb_changes]}
     ],
-    {ok, {{one_for_one,10,1}, Children}}.
+    {ok, {{rest_for_one,10,1}, Children}}.
```
Why rest_for_one and not one_for_one?

nickva replied:
It is rest_for_one because there are some dependencies between components. One case is scheduler_sup and scheduler (scheduler calls scheduler_sup). Others would be clustering and multidb_change -- when a replication doc is processed it calls clustering gen_server at some point to see if cluster is stable, so want clustering to be up before. multidb_changes is also last in list, it can restart (and will be restarted during rescan routinely) and it shouldn't affect processes above it.

sagelywizard replied:
Hm. What're the dependencies exactly? Using a `rest_for_one` is just a little opaque as far as communicating those intentions. Are each of these really dependent on all the prior procs existing? If not, it might be cleaner to have a couple `one_for_all` supervisors and an `one_for_one` sup.

nickva replied:
Looking more in depth at dependencies:
- `_scheduler` needs `_scheduler_sup`
- replications spawned from `_multidb_changes` need `_scheduler`, event manager, and `_clustering`. 
- If `_clustering` fails we want `_multidb_changes` to restart.  
- If `_multidb_changes` fails, other children don't need to restart.  
- If `_scheduler` crashes, we want to `multidb_changes` to restart so it can rescan and re-add the jobs.
- If `_clustering` crashes we'd have to restart scheduler
- If event manager crashes, replications running in scheduler need to restart. So it should move above the scheduler.

sagelywizard commented on src/couch_replicator.erl:
```diff
@@ -103,935 +68,272 @@ replicate(PostBody, Ctx) ->
+clean_up_replications(DbName) ->
+    RepIds = find_jobs_by_dbname(DbName),
+    [couch_replicator_scheduler:remove_job(RepId) || RepId <- RepIds],
+    ok.
```
Another nit: You can replace the above two lines with `lists:foreach(fun couch_replicator_scheduler:remove_job/1, RepIds).`

nickva replied:
Good find!

nickva committed:
Split couch_replicator.erl into 2 modules.

`couch_replicator_doc_processor`
  - Handles multi-db changes callbacks so it starts replications
    defined in _replicator docs.

`couch_replicator`
  - Handles `_replicate` http endpoint and creating local
    `_replicator` db.

nickva committed:
Update couch_replicator_sup. Move clustering and event manager to the top.

 - If `_clustering` crashes we'd have to restart scheduler. Replication used
clustering module to decide if they should still run on current. Node. Move
clustering before scheduler in child order. If scheduler crashes, clustering
doesn't need to be restarted.

 - If event manager crashes, replications running in scheduler need to restart.
So it should move above the scheduler. If scheduler crashes event manager can
be still left running.

Other dependencies:

 * `_scheduler` needs `_scheduler_sup`
 * Replications spawned from `_multidb_changes` need `_scheduler`,
   event manager, and `_clustering`.
 * If `_clustering` fails we want `_multidb_changes` to restart.
 * If `_multidb_changes` fails, other children don't need to restart.
 * If `_scheduler` crashes, we want to `multidb_changes` to restart so it can
   rescan and re-add the jobs.

davisp commented:
I'm not against this on the face of it, but I'm confused on the motivation. If we're re-writing the replicator then I'd rather just focus on this sort of clean up in the rewrite. If we're planning on refactoring the current replicator until its no longer full of terrible then I'd be more inclined to read this more thoroughly.

However, in the first case its a bit concerning to be adding a lot of churn when we're just going to be abandoning it in hopefully a short time frame.

nickva commented:
@davisp  This is part of the rewrite. We were thinking it would be part of what we called "phase 1" -- adding a scheduler to avoid running all replications at once, and coming up with  a less complicated replication manager and supervision hierarchy.  So we split some of the functionality which replication manager did into separate processes (cluster management, replication db doc change handling, etc).

nickva committed:
Improve clustering module & rescan handling

Now there is a grace period during startup where nodes are expected to be added
to the cluster. During that period a shorter quiet time applies (5 seconds by
default). This allow replicator change feeds to start quicker, instead of
waiting for more than a minute.

Instead of directly manipulating couch_multidb_changes process from the
supervisor. Add a separate process (couch_replicator_db_change) which handles
rescannig in a cleaner way.

Rescanning is triggered by an event from clustering module. Events are sent
through the `replication` gen_event (`couch_replicator_notifier` module).

In previous commit forgot to actually create the local `_replicator` db. This
commit fixes that.

Also added more dialyzer types.

Eunit tests (171 still pass). Replications checked via `rep` script still work.

```
In [1]: rep.replicate_1_to_n_then_check_replication(100)
creating rdyno_src_0001
...
 > function wait_to_trigger succeded after 10.509 +/- 10.0  sec.
all replication documents triggered
>>> update cycle 0  <<<
 > waiting for target  1
 > retrying function wait_till_dbs_equal
 > function wait_till_dbs_equal succeded after 6.414 +/- 1.0  sec.
...
> waiting to propagate changes from  1 to 100  : 48.636 sec.
 ```

rnewson commented on src/couch_multidb_changes.erl:
```diff
@@ -145,7 +145,7 @@ handle_cast({resume_scan, DbName}, #state{pids=Pids, tid=Ets} = State) ->
 handle_info({'EXIT', From, normal}, #state{scanner = From} = State) ->
-    couch_log:info("multidb_changes ~p scanner pid exited ~p",[State#state.suffix, From]),
+    couch_log:debug("multidb_changes ~p scanner pid exited ~p",[State#state.suffix, From]),
     {noreply, State#state{scanner=nil}};
```
maybe it'll show up later, but we usually restart such things if we're monitoring their exit.

nickva replied:
In this case it is a normal exit and we don't restart scanner. It is a debug message saying "scanner has finished finding all local db shards and exited". From then on event handler on local db shards will handle updates / new db shards. Below we handle the case of abnormal exit of scanner and stop the whole multidb_changes process  (in order to trigger a restart through supervisor).

rnewson replied:
ok, then I think a notice level statement that confirms normal completion would be better.

rnewson commented on src/couch_replicator.erl:
```diff
@@ -191,847 +97,47 @@ wait_for_result(RepId) ->
-        Error
+-spec cancel_replication(rep_id()) ->
+    {ok, {cancelled, binary()}} | {error, not_found}.
+cancel_replication({BasedId, Extension} = RepId) ->
```
pass the Rep in and pass Rep#rep.id to remove_job, then we can ditch the ETS table.

rnewson commented on src/couch_replicator_clustering.erl:
```diff
@@ -42,94 +59,79 @@ start_link() ->
+% is considered to be unstable i.e. it has changed recently, or returns
+% node() which is considered to be the owner.
 %
 -spec owner(Dbname :: binary(), DocId :: binary() | null) ->
```
I'd like to simplify this, have a single owner function that takes a #rep{} or term(). In the case of unstable, we can return `nonode`, since we would do the same thing as if the job was owned by a foreign node (not start it if asked, stop it on next checkpoint).

The contract can be owner(#rep{} | term()) -> node() for all cases, we'll just sometimes return a node that is none of the running nodes.

nickva replied:
Deciding on ownership based on `{dbname, docid}` means we don't make potentially a large number of external http requests to get filters (for js filters)  to compute replication ids. Currently we only look at document  info, then decide to skip it if it doesn't belong to the node. But if we disregard that corner case, having ownership decided based on #rep would be more elegant.

The other part involves trickery of what the return values mean. In this pr, ets table from replication manager is gone, so `nonode` should never be returned. But `no_owner` is a return value that means "there is no doc_id for this replication" i.e. this is `_replicate` replication whose ownership has already been determine on creation and after that it is not shuffled around.

rnewson commented on src/couch_replicator_scheduler_job.erl:
```diff
@@ -332,9 +334,9 @@ handle_cast({db_compacted, DbName},
-    case couch_replicator_manager:continue(Rep) of
-    {true, _} ->
+    #rep_state{rep_details = #rep{db_name = DbName, doc_id = DocId} = Rep} = State,
+    case couch_replicator_clustering:owner(DbName, DocId) of
```
let's pass Rep here.

nickva replied:
There is note above related to this. Switching ownership to be based on #rep{} might have performance implications for filter functions.

rnewson replied:
how?

nickva replied:
I think I got confused thinking to have a #rep{} we'd have to go through computing the replication_id

rnewson commented:
+1
